src/db.c

```diff
@@ -4128,206 +4155,1718 @@ db_speaker_clear_all(void)
+  if (id == 0)
+    return;
+
+  query = sqlite3_mprintf(Q_TMPL, artist, album, id);
+  if (!query)
+    {
+      DPRINTF(E_LOG, L_DB, "Out of memory for query string\n");
+
+      return;
+    }
+
+  db_query_run(query, 1, 0);
+  queue_inc_version_and_notify();
 
-  return db_query_run(query, 1, 0);
 #undef Q_TMPL
 }
 
-int
-db_watch_delete_bycookie(uint32_t cookie)
-{
-#define Q_TMPL "DELETE FROM inotify WHERE cookie = %" PRIi64 ";"
+static int
+queue_add_mediafileinfo(struct db_media_file_info *dbmfi, int pos, int shuffle_pos)
+{
+#define Q_TMPL "INSERT INTO queue "							\
+		    "(id, file_id, song_length, data_kind, media_kind, "		\
+		    "pos, shuffle_pos, path, virtual_path, title, "			\
+		    "artist, album_artist, album, genre, songalbumid, "			\
+		    "time_modified, artist_sort, album_sort, album_artist_sort, year, "	\
+		    "track, disc)" 							\
+		"VALUES"                                           			\
+		    "(NULL, %s, %s, %s, %s, "						\
+		    "%d, %d, %Q, %Q, %Q, "						\
+		    "%Q, %Q, %Q, %Q, %s, "						\
+		    "%s, %Q, %Q, %Q, %s, "						\
+		    "%s, %s);"
+
   char *query;
+  int ret;
 
-  if (cookie == 0)
-    return -1;
+  query = sqlite3_mprintf(Q_TMPL,
+			  dbmfi->id, dbmfi->song_length, dbmfi->data_kind, dbmfi->media_kind,
+			  pos, pos, dbmfi->path, dbmfi->virtual_path, dbmfi->title,
+			  dbmfi->artist, dbmfi->album_artist, dbmfi->album, dbmfi->genre, dbmfi->songalbumid,
+			  dbmfi->time_modified, dbmfi->artist_sort, dbmfi->album_sort, dbmfi->album_artist_sort, dbmfi->year,
+			  dbmfi->track, dbmfi->disc);
+  ret = db_query_run(query, 1, 0);
 
-  query = sqlite3_mprintf(Q_TMPL, (int64_t)cookie);
+  return ret;
 
-  return db_query_run(query, 1, 0);
 #undef Q_TMPL
 }
 
-static int
-db_watch_get_byquery(struct watch_info *wi, char *query)
+/*
+ * Adds the files matching the given query to the queue after the item with the given item id
+ *
+ * The files table is queried with the given parameters and all found files are added after the
+ * item with the given item id to the "normal" queue. They are appended to end of the shuffled queue
+ * (assuming that the shuffled queue will get reshuffled after adding new items).
+ *
+ * The function returns -1 on failure (e. g. error reading from database) and if the given item id
+ * does not exist. It wraps all database access in a transaction and performs a rollback if an error
+ * occurs, leaving the queue in a consistent state.
+ *
+ * @param qp Query parameters for the files table
+ * @param item_id Files are added after item with this id
+ * @return 0 on success, -1 on failure
+ */
+int
+db_queue_add_by_queryafteritemid(struct query_params *qp, uint32_t item_id)
 {
-  sqlite3_stmt *stmt;
-  char **strval;
-  char *cval;
-  uint32_t *ival;
-  int64_t cookie;
-  int ncols;
-  int i;
+  struct db_media_file_info dbmfi;
+  char *query;
+  int shuffle_pos;
+  int pos;
   int ret;
 
-  DPRINTF(E_DBG, L_DB, "Running query '%s'\n", query);
+  db_transaction_begin();
```

Comment:
It is necessary to wrap the reads and writes in a single transaction to guarantee that a parallel update does not change the values between reading and writing. This may lead to an inconsistent queue, with missing or duplicate entries for one position.